system/lib/libc/musl/src/signal/setitimer.c

```diff
@@ -37,11 +37,15 @@ void _emscripten_timeout(int which, double now)
 		signum = SIGPROF;
 	else if (which == ITIMER_VIRTUAL)
 		signum = SIGVTALRM;
-	int next_timeout = current_intervals_ms[which];
-	if (next_timeout)
-		current_timeout_ms[which] = now + next_timeout;
-	else
+	double next_timeout = 0.0;
+	if (current_intervals_ms[which]) {
+		// The next alarm is due 'interval' ms after the previous one, which may
+		// be sooner than 'interval' ms from now (if this alarm was delayed)
+		current_timeout_ms[which] += current_intervals_ms[which];
+		next_timeout = current_timeout_ms[which] - now;
```

Comment:
oh, yes indeed; In that case an event(s) is lost. I guess in that case we still want to fire the next event at the time when an event would have been scheduled had there been no delays (i.e. less than 'interval' ms in the future).

Reply:
Why a multiple in the latest code? It seems like if the interval is large, then just missing the time would lead to another large delay. Or is there a constraint on the time between the events and not just their absolute times?

Second, I'm worried about negative numbers here. If `now` is 1000, `timeout` is 1002, and `interval` is 1 then we end up with `uint64_t(-1)`, I think? (If so it might be simpler to separate the cases with an `if (now < timeout)`?)

Reply:
(unresolving, I think my comment might have raced with your resolving it)

Reply:
The multiple is for the case where we have a big delay and miss multiple alarms.
A simple extension to my original code would be `while (current_timeout_ms < now) current_timeout_ms += current_interval_ms;` But if time goes very far forward (e.g. if the machine goes to sleep for a long time and the interval is short), then the loop could take a long time. So the current code just tries to do that by finding how many intervals we need to go forward by. The common low-load case is where `now` is just slightly later than `current_timeout_ms` in which case the division will truncate to 0 and `intervals` will be 1. The target time should always be less than `current_intervals_ms` ahead of `now`.

Regarding negatives:
`now` Should never be `< current_timeout`. We check that at the callsite in `emscripten_check_timers`. But yeah, I guess speaking of time doing unexpected things we should probably have a graceful failure/reset in case time goes backwards.

Reply:
I see. But do we need to preserve the intervals between events precisely (which is what I think the loop helps with)? If not, then this is simpler I think. And I think we don't need to be precise because timers can't be, anyhow.

That is, I think maybe we can just set a timer for the interval?

Reply:
Let me back up: the whole point of the new logic here is to preserve the regularity of events better, especially in the case where events are delayed due to system load or whatever. It's not guaranteed to be precise, but the current simple logic (of always just scheduling 'interval' ms in the future) leads to skew that accumulates over time; this happens even when load is light, since the events are never delivered precisely on time. 

This is the main reason the test is currently failing: the test expects to receive a certain number of events in a certain period of time, and it doesn't when there is load. IMHO this is the part of the point of an interval timer; if you want the behavior of getting a minimum delay after the callback, you can always chain individual timers from the callback itself.

If you don't agree, then we should just not change the implementation, and instead change the test so that it expects different behavior.

Reply:
I see, thanks. I guess I'm not sure what the real-world requirements are here, but aiming for the right frequency seems the safest thing, I agree.

I think the problem of negative numbers here remains, though? With just fixing that I think we'd be ok. And fixing it by doing the minimum delay possible seems best, as the way to get closest to the right frequency?

Reply:
Yeah, negative numbers can happen if time goes backwards and IIUC we sometimes have date.now  which is not monotonic. I'll add a check here and we can separately discuss that.